File: modules/core/src/main/java/org/apache/ignite/services/ServiceCallContext.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.services;

import org.apache.ignite.lang.IgniteExperimental;

/**
 * Service call context.
 * <p>
 * This context is implicitly passed to the service and can be retrieved inside the service using {@link
 * ServiceContext#currentCallContext()}. It is accessible only from the local thread during the execution of a service
 * method.
 * <p>
 * Use {@link ServiceCallContext#builder()} to create the context builder.
 * <p>
 * <b>Note</b>: passing the context to the service may lead to performance overhead, so it should only be used for
 * "middleware" tasks.
 * <p>
 * Usage example:
 * <pre name="code" class="java">
 *
 * // Service implementation.
 * class HelloServiceImpl implements HelloService {
 *     &#64;ServiceContextResource
 *     ServiceContext ctx;
 *
 *     public String call(Strig msg) {
 *         return msg + ctx.currentCallContext().attribute("user");
 *     }
 *     ...
 * }
 * ...
 *
 * // Call this service with context.
 * ServiceCallContext callCtx = ServiceCallContext.builder().put("user", "John").build();
 * HelloService helloSvc = ignite.services().serviceProxy("hello-service", HelloService.class, false, callCtx, 0);
 * // Print "Hello John".
 * System.out.println( helloSvc.call("Hello ") );
 * </pre>
 *
 * @see ServiceContext
 * @see ServiceCallContextBuilder
 * @see ServiceCallInterceptor
 */
@IgniteExperimental
public interface ServiceCallContext {
    /**
     * Create a context builder.
     *
     * @return Context builder.
     */
    public static ServiceCallContextBuilder builder() {
        return new ServiceCallContextBuilder();
    }

    /**
     * Get the string attribute.
     *
     * @param name Attribute name.
     * @return String attribute value.
     */
    public String attribute(String name);

    /**
     * Get the binary attribute.
     *
     * @param name Attribute name.
     * @return Binary attribute value.
     */
    public byte[] binaryAttribute(String name);
}

File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Services/ServiceCallContext.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache.Ignite.Core.Impl.Services
{
    using System.Collections;
    using Apache.Ignite.Core.Impl.Common;
    using Apache.Ignite.Core.Services;

    /// <summary>
    /// Service call context implementation.
    /// </summary>
    internal class ServiceCallContext : IServiceCallContext
    {
        /** Context attributes. */
        private IDictionary _attrs;

        /// <summary>
        /// Constructs context from dictionary.
        /// </summary>
        /// <param name="attrs">Context attributes.</param>
        internal ServiceCallContext(IDictionary attrs)
        {
            IgniteArgumentCheck.NotNull(attrs, "attrs");

            _attrs = attrs;
        }

        /** <inheritDoc /> */
        public string GetAttribute(string name)
        {
            return (string) _attrs[name];
        }

        /** <inheritDoc /> */
        public byte[] GetBinaryAttribute(string name)
        {
            return (byte[]) _attrs[name];
        }
        
        /// <summary>
        /// Gets call context attributes.
        /// </summary>
        /// <returns>Service call context attributes.</returns>
        internal IDictionary Values()
        {
            return _attrs;
        }
    }
}